File: 算法/2. 两数相加/AddTwoNumbers.go
package ___两数相加

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	node := &ListNode{}
	cur := node
	flag := 0
	for l1 != nil || l2 != nil {
		var m = 0
		var n = 0
		if l1 != nil {
			m = l1.Val
		}
		if l2 != nil {
			n = l2.Val
		}
		sum := m + n + flag
		flag = sum / 10
		curValue := sum % 10
		curNode := &ListNode{
			Val: curValue,
		}
		cur.Next = curNode
		cur = cur.Next
		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}
	}
	// 最后如果还有进位未处理需要加一
	if flag == 1 {
		cur.Next = &ListNode{
			Val: flag,
		}
	}
	return node.Next
}

File: 算法/2. 两数相加/AddTwoNumbers.java

public class AddTwoNumbers {


    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    class Solution {
        public ListNode addTwoNumbers1(ListNode l1, ListNode l2) {
            ListNode node = new ListNode();
            ListNode cur = node;
            int flag = 0;
            while (l1 != null || l2 != null) {
                int m = l1 == null ? 0 : l1.val;
                int n = l2 == null ? 0 : l2.val;
                int sum = m + n + flag;
                flag = sum / 10;
                int nowNum = sum % 10;
                cur.next = new ListNode(nowNum);
                cur = cur.next;
                if (l1 != null) {
                    l1 = l1.next;
                }
                if (l2 != null) {
                    l2 = l2.next;
                }
            }
            if (flag == 1) {
                cur.next = new ListNode(1);
            }
            return node.next;
        }

        public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
            ListNode result = new ListNode();
            ListNode cur = result;
            int flag = 0;
            while (l1 != null || l2 != null) {
                int m = l1 == null ? 0 : l1.val;
                int n = l2 == null ? 0 : l2.val;
                int sum = m + n + flag;
                int curValue = sum % 10;
                flag = sum / 10;
                cur.next = new ListNode(curValue);
                cur = cur.next;
                if (l1 != null) {
                    l1 = l1.next;
                }
                if (l2 != null) {
                    l2 = l2.next;
                }
            }
            if (flag != 0) {
                cur.next = new ListNode(flag);
            }
            return result.next;
        }
    }

}
